engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/UpdateBy.java

```diff
@@ -642,68 +690,118 @@ private void prepareForParallelPopulation(final Runnable resumeAction) {
                             }
                         }
-                    }, resumeAction, this::onError);
+                    }, onParallelPopulationComplete, this::onError);
         }
 
-        private void processWindowOperators(final int winIdx, final Runnable resumeAction) {
+        /**
+         * Process the operators for a given window in a serial manner. For efficiency, this function organizes the
+         * operators into sets of operators that share input sources and that can be computed together efficiently. It
+         * also arranges these sets of operators in an order that (hopefully) minimizes the memory footprint of the
+         * cached operator input columns.
+         *
+         * Before each operator set is processed, the sources for the input columns are cached. After the set is
+         * processed, the cached sources are released if they will not be used by following operators.
+         */
+        private void processWindowOperators(
+                final int winIdx,
+                final int maxAffectedChunkSize,
+                final int maxInfluencerChunkSize,
+                final Runnable onProcessWindowOperatorsComplete,
+                final Consumer<Exception> onProcessWindowOperatorsError) {
             final UpdateByWindow win = windows[winIdx];
 
-            // Arrange the dirty operators to increase the chance that the input caches can be released early
+            // Order the dirty operators to increase the chance that the input caches can be released early
             final Integer[] dirtyOperators = ArrayUtils.toObject(dirtyWindowOperators[winIdx].stream().toArray());
-            Arrays.sort(dirtyOperators, Comparator.comparingInt(o -> win.operatorInputSourceSlots[o][0]));
+            Arrays.sort(dirtyOperators,
+                    Comparator.comparingInt(o -> win.operatorInputSourceSlots[(int) o][0])
+                            .thenComparingInt(o -> win.operatorInputSourceSlots[(int) o].length < 2 ? -1
+                                    : win.operatorInputSourceSlots[(int) o][1]));
+
+            final List<int[]> operatorBins = new ArrayList<>(dirtyOperators.length);
+            final TIntArrayList opList = new TIntArrayList();
+            opList.add(dirtyOperators[0]);
+            int lastOpIdx = dirtyOperators[0];
+            for (int ii = 1; ii < dirtyOperators.length; ii++) {
+                final int opIdx = dirtyOperators[ii];
+                if (Arrays.equals(win.operatorInputSourceSlots[opIdx], win.operatorInputSourceSlots[lastOpIdx])) {
+                    opList.add(opIdx);
+                } else {
+                    operatorBins.add(opList.toArray());
+                    opList.clear(dirtyOperators.length);
+                    opList.add(opIdx);
+                }
+                lastOpIdx = opIdx;
+            }
+            operatorBins.add(opList.toArray());
+            final int[][] dirtyOperatorBins = operatorBins.toArray(int[][]::new);
```

Comment:
You could consider just using lists instead of arrays.

Reply:
These get passed and re-used many times by subsequent functions, decided it's easier to convert once and reuse.